PolyEngine/Engine/Src/ECS/ComponentIterator.hpp

```diff
@@ -0,0 +1,109 @@
+			virtual Entity* get() const = 0;
+			virtual void increment() = 0;
+			virtual bool isValid() const = 0;
+	};
+
+	template<typename PrimaryComponent, typename... SecondaryComponents>
+	class ComponentIterator : public BaseObject<>,
+							public std::iterator<std::forward_iterator_tag, std::tuple<typename std::add_pointer<PrimaryComponent>::type, typename std::add_pointer<SecondaryComponents>::type...>>	
+	{
+		public:
+			explicit ComponentIterator(std::unique_ptr<IEntityIteratorHelper> iter) : Iter(std::move(iter))
+			{
+				UpdateCache();
+			}
+			bool operator==(const  ComponentIterator& rhs) const { return Iter.get()->get() == rhs.Get()->get(); } //they never  trigger underlying overload, pointer comparison ongoing
+			bool operator!=(const  ComponentIterator& rhs) const { return !(Iter.get()->get() == rhs.Get()->get()); }
+
+			std::tuple<typename std::add_pointer<PrimaryComponent>::type, typename std::add_pointer<SecondaryComponents>::type... > operator*() const //canot be const if  change cache
+			{
+				return GetCache();
+			}
+			std::tuple<typename std::add_pointer<PrimaryComponent>::type, typename std::add_pointer<SecondaryComponents>::type... > operator->() const
+			{
+				return GetCache();
+			}
+
+			ComponentIterator& operator++() { Increment(); return *this; }
+			ComponentIterator operator++(int) { ComponentIterator ret(Iter); Increment(); return ret; } //test for double incrementing etc
+
+			IEntityIteratorHelper* Get() const
+			{
+				return Iter.get();
+			}
+		private:
+			void Increment()
+			{
+				//ASSERTE(Iter.get()->isValid(), "Next Iterator is not valid!");	//currently not working as we never set it to nullptr?
+				Iter.get()->increment();	
+				bCacheValid = false;
+			}
+
+			void UpdateCache() const 
+			{
+				Entity* ent = Iter.get()->get();
+				PrimaryComponent* primary = ent->GetComponent<PrimaryComponent>();
+				if (!primary)
+				{
+					bCacheValid = false;
+					return;
+				}
+				if constexpr (!Trait::IsVariadicEmpty<SecondaryComponents...>::value)
+				{
+					Cache = std::make_tuple(primary, primary->template GetSibling<SecondaryComponents>()...);
+					bCacheValid = true;
+					return;
+				}
+				else
+				{
+					Cache = std::make_tuple(primary);
+				}
+
+				bCacheValid = true;
+			}
+
+			std::tuple<typename std::add_pointer<PrimaryComponent>::type, typename std::add_pointer<SecondaryComponents>::type... > GetCache() const
+			{
+				if (!bCacheValid)
+					UpdateCache();
+				// Entity* ent = Iter.get()->get();
+				// if (!HasComponents<PrimaryComponent, SecondaryComponents...>(ent))
+				// 	ASSERTE(false, "Updated cache does not contain all needed components!"); //@fixme: celeborth, when put in asserte clause it fails (passing to macro fails template deduction?)
+				return Cache;
+			}
+
+			//------------------------------------------------------------------------------
+			template<int zero = 0>
+			bool HasComponents(const Entity* entity) const { return true; }
```

Comment:
You never use this method. Do we ensure that the secondary components are never null? I don't think we do. At least not in this piece of code, and it should be handled by the ComponentIterator, not the IteratorHelper.

.